src/diagnostics/Diagnostics.jl

```diff
@@ -0,0 +1,634 @@
+    output_every = sd_time.output_every * Δt
+
+    ScheduledDiagnosticTime(;
+        sd_time.variable,
+        output_every,
+        sd_time.output_writer,
+        sd_time.reduction_time_func,
+        sd_time.reduction_space_func,
+        compute_every,
+        sd_time.pre_output_hook!,
+    )
+end
+
+# We provide also a companion constructor for ScheduledDiagnosticIterations which returns
+# itself (without copy) when called with a timestep.
+#
+# This is so that we can assume that
+# ScheduledDiagnosticIterations(ScheduledDiagnostic{Time, Iterations}, Δt)
+# always returns a valid ScheduledDiagnosticIterations
+ScheduledDiagnosticIterations(
+    sd::ScheduledDiagnosticIterations,
+    _Δt::T,
+) where {T} = sd
+
+
+# We define all the known identities in reduction_identities.jl
+include("reduction_identities.jl")
+
+# Helper functions
+include("diagnostics_utils.jl")
+
+"""
+    get_callbacks_from_diagnostics(diagnostics, storage, counters)
+
+
+Translate a list of diagnostics into a list of callbacks.
+
+Positional arguments
+=====================
+
+- `diagnostics`: List of `ScheduledDiagnosticIterations` that have to be converted to
+                 callbacks. We want to have `ScheduledDiagnosticIterations` here so that we
+                 can define callbacks that occur at the end of every N integration steps.
+
+- `storage`: Dictionary that maps a given `ScheduledDiagnosticIterations` to a potentially
+             pre-allocated area of memory where to accumulate/save results.
+
+- `counters`: Dictionary that maps a given `ScheduledDiagnosticIterations` to the counter
+              that tracks how many times the given diagnostics was computed from the last
+              time it was output to disk.
+
+"""
+function get_callbacks_from_diagnostics(diagnostics, storage, counters)
+    # We have two types of callbacks: to compute and accumulate diagnostics, and to dump
+    # them to disk. Note that our callbacks do not contain any branching
+
+    # storage is used to pre-allocate memory and to accumulate partial results for those
+    # diagnostics that perform reductions.
+
+    callbacks = Any[]
+
+    for diag in diagnostics
+        variable = diag.variable
+        isa_reduction = !isnothing(diag.reduction_time_func)
+
+        # reduction is used below. If we are not given a reduction_time_func, we just want
+        # to move the computed quantity to its storage (so, we return the second argument,
+        # which that will be the newly computed one). If we have a reduction, we apply it
+        # point-wise
+        reduction = isa_reduction ? diag.reduction_time_func : (_, y) -> y
+
+        # If we have a reduction, we have to reset the accumulator to its neutral state. (If
+        # we don't have a reduction, we don't have to do anything)
+        #
+        # ClimaAtmos defines methods for identity_of_reduction for standard
+        # reduction_time_func in reduction_identities.jl
+        reset_accumulator! =
```

Comment:
I'm a bit wary of having too many closures:
- they make stack traces harder to read (anonymous functions don't get nice names, names get mangled)
- they can capture variables unexpectedly, making it harder to figure out what is in scope

As an alternative, I would propose defining a function at top-level
```julia
function reset_accumulator!(diag_storage, diag::ScheduledDiagnostic)
    if !isnothing(diag.reduction_time_func) # this check should be inferred at compile time, so function would become a no-op
        # identity_of_reduction works by dispatching over operation
        identity = identity_of_reduction(diag.reduction_time_func)
        float_type = eltype(diag_storage)
        identity_ft = convert(float_type, identity)
        diag_storage .= identity_ft
    end
end
```

Reply:
And similarly
```julia
function apply_reduction!(diag_storage, reductfn, value)
   diag_storage .= reductfn.(diag_storage, value)
end
function apply_reduction!(diag_storage, reductfn::Nothing, value)
   diag_storage .= value
end
```